--- /dev/null
+++ src/bazel_runfiles.cc
@@ -0,0 +1,53 @@
+#include "tools/cpp/runfiles/runfiles.h"
+
+#include <cstring>
+#include <string>
+
+using bazel::tools::cpp::runfiles::Runfiles;
+
+struct BazelRunfiles {
+	std::unique_ptr<Runfiles> ptr;
+};
+
+extern "C" {
+
+static struct BazelRunfiles *bazel_runfiles = NULL;
+
+void bazel_runfiles_init(const char *argv0_cstr) {
+	if (bazel_runfiles != nullptr) {
+		return;
+	}
+
+	Runfiles *runfiles;
+	std::string argv0(argv0_cstr);
+	std::string source_repository(BAZEL_CURRENT_REPOSITORY);
+	std::string error;
+	runfiles = Runfiles::Create(argv0, source_repository, &error);
+	if (runfiles == nullptr) {
+		return;
+	}
+
+	bazel_runfiles = new BazelRunfiles;
+	bazel_runfiles->ptr.reset(runfiles);
+}
+
+char *bazel_runfiles_bison_pkgdatadir() {
+	if (bazel_runfiles == nullptr) {
+		return nullptr;
+	}
+
+	std::string path = std::string(
+		BAZEL_CURRENT_REPOSITORY "/data/m4sugar/m4sugar.m4"
+	);
+	std::string result = bazel_runfiles->ptr->Rlocation(path);
+	if (result.empty()) {
+		return nullptr;
+	}
+
+	char *datadir = strdup(result.c_str());
+	/* "data/m4sugar/m4sugar.m4" => "data" */
+	datadir[result.size() - 19] = '\x00';
+	return datadir;
+}
+
+}
